### ISSUE TITLE
ES6 imports in dist files creates issues with server side rendering

### ISSUE THREAD

**Issue by silviogutierrez** [ionic-team/ionic-native#3279]
**I'm submitting a ...**  (check one with "x")
[x] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or https://ionicworldwide.herokuapp.com/

**Current behavior:**
Obviously we don't want to use any of this hardware when server side rendering. But if we have a unified code base, if you do `import {BarcodeScanner} from '@ionic-native/http'`, for example, you're going to run into node issues when you visit any page. Imports are static, so everything is processed as long as the require tree hits this file.

**Expected behavior:**
End-distribution files should stick to CommonJS imports and exports, along with declaration files.

**Steps to reproduce:**
1. Have server.ts as an entry point to your SSR logic.
2. In that, import `BarcodeScanner`
3. Try to visit the server, even if you don't use `BarcodeScanner` (and of course, you won't, it's the server)`

**Related code:**
Troublesome code, affecting all of the plugins I use, is in the index.js files. Example:
`node_modules/@ionic-native/barcode-scanner/index.js` has

`import { IonicNativePlugin, cordova } from '@ionic-native/core';`

which will cause `SyntaxError: Unexpected token import`

**Other information:**
The fix is quite easy. When publishing to NPM, compile these files down to ES5.

Current workaround I use is messy, but I centralized all my hardware plugins into a single `hardware.ts` file that automagically only imports when running on the client, while still keeping type support:

```typescript
const LaunchReviewImport = async () => (await import('@ionic-native/launch-review')).LaunchReview;
const BarcodeScannerImport = async () => (await import('@ionic-native/barcode-scanner')).BarcodeScanner;
const HTTPImport = async () => (await import('@ionic-native/http')).HTTP;

type Unpromise<T extends Promise<any>> = T extends Promise<infer U> ? U : never;


export const LaunchReview: Unpromise<ReturnType<typeof LaunchReviewImport>> = (typeof window === 'undefined') ? (null as any) : require('@ionic-native/launch-review').LaunchReview;
export const BarcodeScanner: Unpromise<ReturnType<typeof BarcodeScannerImport>> = (typeof window === 'undefined') ? (null as any) : require('@ionic-native/barcode-scanner').BarcodeScanner;
export const HTTP: Unpromise<ReturnType<typeof HTTPImport>> = (typeof window === 'undefined') ? (null as any) : require('@ionic-native/http').HTTP;
```

Again, the nulls don't matter here, because all of this hardware is used on event handlers, which by definition only run on the client.

**Comment by mhartington** [comment]
Can you isolate this in a small sample app? Testing this in an angular universal (their v9 RC) and Im not seeing any issues.

**Comment by silviogutierrez** [comment]
I can give it a try, but on your end: which server page are you rendering? Make sure it's a page or route that hits the right import cycle. Meaning, one that imports `hardware.ts` for example.

**Comment by laurentgoudet** [comment]
I can confirm this is a valid issue, as I'm hitting it as well.

> Can you isolate this in a small sample app? Testing this in an angular universal (their v9 RC) and Im not seeing any issues.

I suspect this is because v9 defaults to bundling the server-side dependencies (which are then transpiled by webpack), whereas pre-v9 these were loaded from `node_modules`.

However, in practice, this cannot be used due to a number of limitations, e.g. i18n support being broken (https://github.com/angular/universal/issues/1454), and has to be reverted by adding    `bundleDependencies": false` in the `options` of the `server` target (in `angular.json`).

> The fix is quite easy. When publishing to NPM, compile these files down to ES5

The files are already transpiled to ES5, the issue is that they use the `es2015` module system, which Node.js does not support (well, it's complicated).

However, only publishing CommonJS modules would break tree-shaking (as per https://webpack.js.org/guides/tree-shaking/), which is one of the reason the Angular Package Format was created in the first place (https://docs.google.com/document/d/1CZC2rcpxffTDfRDs6p1cfbmKNLA6x5O-NtkJglDaBVs/preview#heading=h.k0mh3o8u5hx).

The TL;DR is that both formats should be provided, ES modules (for tree-shaking on the browser side - ideally flattened) & CommonJS bundles (loaded by Node.js): [ng-packagr](https://github.com/ng-packagr/ng-packagr) does that well for simple libraries, but not sure how practical it would be in the case of the `ionic-native` build.
